lib/reek/configuration/app_configuration.rb

```diff
@@ -10,135 +14,95 @@ module Configuration
     #
     # @api private
     class AppConfiguration
+      include ConfigurationValidator
       EXCLUDE_PATHS_KEY = 'exclude_paths'
-      attr_reader :exclude_paths, :default_directive, :directory_directives
+      private_attr_writer :directory_directives, :default_directive, :excluded_paths
 
-      # Given this configuration file:
+      # Instantiate a configuration via given path.
       #
-      # ---
-      # IrresponsibleModule:
-      #   enabled: false
-      # "app/helpers":
-      #   UtilityFunction:
-      #     enabled: false
-      # exclude_paths:
-      #   - "app/controllers"
+      # @param path [Pathname] the path to the config file
       #
-      # this would result in the following configuration:
-      #
-      # exclude_paths = [ Pathname('spec/samples/two_smelly_files') ]
-      # default_directive = { Reek::Smells::IrresponsibleModule => { "enabled" => false } }
-      # directory_directives = { Pathname("spec/samples/three_clean_files/") =>
-      #                          { Reek::Smells::UtilityFunction => { "enabled" => false } } }
+      # @return [AppConfiguration]
+      def self.from_path(path: nil)
+        allocate.tap do |instance|
+          instance.instance_eval { find_and_load(path: path) }
+        end
+      end
+
+      # Instantiate a configuration by passing everything in.
       #
+      # @param directory_directives [Hash] for instance:
+      #   { Pathname("spec/samples/three_clean_files/") =>
+      #     { Reek::Smells::UtilityFunction => { "enabled" => false } } }
+      # @param default_directive [Hash] for instance:
+      #   { Reek::Smells::IrresponsibleModule => { "enabled" => false } }
+      # @param excluded_paths [Array] for instance:
+      #   [ Pathname('spec/samples/two_smelly_files') ]
       #
-      # @param options [OpenStruct]
-      #   e.g.: #<OpenStruct report_format =: text,
-      #                      location_format =: numbers,
-      #                      colored = true,
-      #                      smells_to_detect = [],
-      #                      config_file = #<Pathname:config/defaults.reek>,
-      #                      argv = [ "lib/reek/spec" ]>
-      def initialize(options = nil)
-        @directory_directives = {}
-        @default_directive    = {}
-        @exclude_paths        = []
+      # @return [AppConfiguration]
+      def self.from_map(directory_directives: {},
+                        default_directive: {},
+                        excluded_paths: [])
+        allocate.tap do |instance|
+          instance.instance_eval do
+            self.directory_directives = directory_directives.extend(DirectoryDirectives)
+            self.default_directive    = default_directive.extend(DefaultDirective)
+            self.excluded_paths       = excluded_paths.extend(ExcludedPaths)
+          end
+        end
+      end
 
-        load options
+      def self.default
+        from_path path: nil
       end
 
+      def initialize(*)
```

Comment:
Wow, I didn't really know about `allocate`. It may be just what I need to fix some long-standing issues with my [GirFFI](https://github.com/mvz/gir_ffi) project.